Fix #1958: Optimize Pin Activity View

bladebunny opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #1958: Flattens view hierarchy leveraging ConstraintLayout.  Uses pre-existing text styles on the relevant text views.
See #1958.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

## Screenshots
![Screenshot_1603239126](https://user-images.githubusercontent.com/1473399/96657669-a8042c80-12f7-11eb-97dc-2cdfbd799bb0.png)
![Screenshot_1603239131](https://user-images.githubusercontent.com/1473399/96657672-a9355980-12f7-11eb-9d6c-185f53cc81fc.png)
![Screenshot_1603239139](https://user-images.githubusercontent.com/1473399/96657674-aa668680-12f7-11eb-8e96-50fc588e6168.png)
![Screenshot_1603239153](https://user-images.githubusercontent.com/1473399/96657677-aaff1d00-12f7-11eb-95c4-322687f43318.png)
![Screenshot_1603239171](https://user-images.githubusercontent.com/1473399/96657678-ab97b380-12f7-11eb-98ad-8b2f07b280a1.png)
![Screenshot_1603239175](https://user-images.githubusercontent.com/1473399/96657679-ab97b380-12f7-11eb-9b44-9623b679016a.png)
![Screenshot_1603239183](https://user-images.githubusercontent.com/1473399/96657680-ac304a00-12f7-11eb-97dc-9172fec9517c.png)

bladebunny committed:
Fixes #1958: Optimize Pin Activity View

Flatten view hierarchy with ConstraintLayout and use standard text styles (Header2 and Subtitle1) instead of inline textview styles.  Small tweaks to margins between controls to be more standard.

rt4914 commented:
@bladebunny Before I do full review of this PR, can you make sure that you update the #1963 PR first. Considering that both of the PRs are similar, once that PR is updated and approved, similar changes should be made in this PR.

For start, you can apply similar changes in this PR too:

https://github.com/oppia/oppia-android/pull/1963#issuecomment-706135392
https://github.com/oppia/oppia-android/pull/1963#pullrequestreview-505603257

bladebunny committed:
Updating margins and padding per code review.

bladebunny commented:
Updated per feedback from #1963.

rt4914 commented on app/src/main/res/layout-land/admin_pin_activity.xml:
```diff
@@ -2,112 +2,97 @@
+    android:background="@color/oppiaBackgroundYellowIvory"
+    android:paddingBottom="108dp">
 
-    <ScrollView
```
Please do not remove the `ScrollView` because considering that the app will be used on low-end devices and it might happen that the entire view does not fix in single screen, we should make this scrollable.

Also, that's one of the reasons why your tests are failing on CI.

<img width="747" alt="Screenshot 2020-10-10 at 10 54 54 AM" src="https://user-images.githubusercontent.com/9396084/95646538-0b67b080-0ae7-11eb-95e6-21df9ace8877.png">


One approach to solve this can be: 
`ScrollView` contains the `ConstraintLayout`

But if that does not work considering that `ScrollView` may create issues with `layout_height = match_parent`
then you can do following:

`FrameLayout` contains `ScrollView` contains the `ConstraintLayout`


Note: We have to flatten the view hierarchy as much as possible but not at the cost of based experience to users.

Thanks.

bladebunny replied:
My bad, I thought there was plenty of room and that the goal to flatten hierarchy was paramount.  Just researched the best practices right for embedded scrollviews with constraintlayout and will update this and my other PR to re-add scrollview.

rt4914 reviewed (changes requested):
Thanks. PTAL (Please take a look)

bladebunny committed:
Adding back scroll support with NestedScrollView

NestedScrollView seems more flexible than ScrollView and achieves the same result.

bladebunny commented:
Re-submitted with scroll support; view hierarchy still somewhat flattened

rt4914 commented:
@bladebunny The UI looks good but the tests are failing because you will need to update the test cases.
mainly you will have to make sure that you use `appendText` instead of `typeText` everything in `AdminPinActivityTest`.

For reference you can check `AddProfileActivityTest` and this PR https://github.com/oppia/oppia-android/pull/1842/files

Also, that being said, I am adding @anandwana001 to this PR to provide more context to you. 
@anandwana001 I made the changes related to appendText in my local but then also the test cases were stuck (meaning the test was not finishing). So can you try and have a look and also guide @bladebunny

anandwana001 commented:
> @bladebunny The UI looks good but the tests are failing because you will need to update the test cases.
> mainly you will have to make sure that you use `appendText` instead of `typeText` everything in `AdminPinActivityTest`.
> 
> For reference you can check `AddProfileActivityTest` and this PR https://github.com/oppia/oppia-android/pull/1842/files
> 
> Also, that being said, I am adding @anandwana001 to this PR to provide more context to you.
> @anandwana001 I made the changes related to appendText in my local but then also the test cases were stuck (meaning the test was not finishing). So can you try and have a look and also guide @bladebunny

Sure, 
currently, the reason for test case failure is because of removing `ScrollView`. Either we can update test cases or UI in order to get them pass.

bladebunny commented:
Sorry about that.  I will fix this morning (US Pacific time).  I wasn't sure how to fix that before but this helps.

bladebunny committed:
Merge branch 'develop' of https://github.com/bladebunny/oppia-android into Fix-1958-Optimize-Pin-Activity-View

bladebunny committed:
Update Tests to appendText instead of typeText

BenHenning commented:
@rt4914 & @anandwana001 deferring to you both since you took a pass on this PR.

@bladebunny FYI I'm also in PST, so if you have questions feel free to reach out. :)

rt4914 commented on app/src/sharedTest/java/org/oppia/android/app/profile/AdminPinActivityTest.kt:
```diff
@@ -742,12 +742,12 @@ class AdminPinActivityTest {
       onView(allOf(withId(R.id.input), isDescendantOfA(withId(R.id.input_confirm_pin)))).perform(
         scrollTo(),
-        typeText("54321"),
+          editTextInputAction.appendText("54321"),
```
Please fix the formatting in this PR as tests lint tests are failing.
<img width="975" alt="Screenshot 2020-10-15 at 4 19 43 AM" src="https://user-images.githubusercontent.com/9396084/96053760-c86a5c00-0e9d-11eb-93b7-0d07e48fe02d.png">

bladebunny replied:
Will do...For some reason I couldn't see the results before - definitely helps to see why it's failing....Will resubmit shortly.

anandwana001 replied:
Check our wiki page here for ktlint setup and fixes - https://github.com/oppia/oppia-android/wiki/Ktlint-Guide

bladebunny replied:
Done

rt4914 replied:
@bladebunny Please do not resolve comments. The person who has started the comment thread should resolve the comment thread. In this case I had started the comment thread and therefore I was suppose to resolve this thread. Let's follow this in future PRs. Thanks.

rt4914 reviewed (changes requested):
@bladebunny Great. Just one change. Also, the tests are failing for `AdminPinActivityTest `, so can you run these tests locally and add screenshot.

bladebunny committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into Fix-1958-Optimize-Pin-Activity-View

bladebunny committed:
Resolve linting issues

bladebunny commented:
It looks RoboElectric is failing now because the base view doesn't inherit from Scroll view or List view - but it is a NestedScrollView - which doesn't inherit from the other view.  It think this assertion/constraint should be updated to include NestedScrollView but I can't figure out how.

anandwana001 commented:
> It looks RoboElectric is failing now because the base view doesn't inherit from Scroll view or List view - but it is a NestedScrollView - which doesn't inherit from the other view. It think this assertion/constraint should be updated to include NestedScrollView but I can't figure out how.

@bladebunny Also, as per the checklist `"Fix #bugnum: ",` could you update the PR title.

anandwana001 commented on app/src/main/res/layout/admin_pin_activity.xml:
```diff
@@ -10,105 +10,96 @@
-      android:id="@+id/scrollViewAdminPin"
+    android:layout_height="match_parent"
+    android:fillViewport="true">
+    <androidx.constraintlayout.widget.ConstraintLayout
```
leave a single line above

bladebunny replied:
Done

anandwana001 replied:
@bladebunny please let the author resolve, you can just comment and the author will check the required changes and mark it resolved.

anandwana001 commented on app/src/main/res/layout/admin_pin_activity.xml:
```diff
@@ -10,105 +10,96 @@
+        android:id="@+id/admin_pin_warning_text"
+        style="@style/Subtitle1"
+        android:layout_height="wrap_content"
+        android:layout_width="match_parent"
```
must be in order. 
Try performing a reformat code here from the menu option in the toolbar.

bladebunny replied:
Done

anandwana001 reviewed (changes requested):
Thanks @bladebunny 
The test cases can pass, please follow the below points:
1. Check this wiki section - https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#tips-to-run-test-cases-in-both-espresso-and-robolectric
2. As we know that this is a sharedTest(run on both espresso and robolectric), so we do have some work around for this NestedScrollView. Please check `AdministratorControlsActivityTest` file, there you will found `nestedScrollView()`, we do require this method in our test file and rather than `scrollTo()` we have to do `nestedScrollView()`. Please take care or teestCoroutineDispatchers also while making changes. 

Let me know if you need any help in this.

bladebunny committed:
Adding nestedScrollTo and fixing xml formatting

bladebunny committed:
Add missing helper functions

anandwana001 commented:
@bladebunny please let the author resolve, you can just comment and the author will check the required changes and mark it resolved. Also, seems like test are working fine now, just for the reference could you please add a screenshot of Android Studio showing that test cases are passing on both espresso and robolectric?

bladebunny commented:
My bad, I thought I was supposed to resolve after addressing their feedback. Isn't it still blocked for their sign off?

rt4914 commented:
> My bad, I thought I was supposed to resolve after addressing their feedback. Isn't it still blocked for their sign off?

@bladebunny Please make sure that you mention the reviewer name and also that we have to take a look otherwise we won't know when to review the PR.

Something like:

@anandwana001 PTAL

anandwana001 commented:
@bladebunny Could you please put the screenshot in the PR description to confirm the UI looks the same as the previous and no tweaks occurred?

bladebunny commented:
@anandwana001 Sure.  I can add screenshot. Any particular requirements?  I know how to take a screenshot but not sure if there are any special instructions.  Thanks.

rt4914 commented:
> @anandwana001 Sure. I can add screenshot. Any particular requirements? I know how to take a screenshot but not sure if there are any special instructions. Thanks.

@bladebunny Just the screenshot should work. Maybe with pin, without pin, etc. under different scenerios.

bladebunny commented:
Screenshots added.

bladebunny commented:
Screenshots added.

anandwana001 reviewed (approved):
LGTM, Thanks @bladebunny 
Just a point to remember, whenever you create a PR which include something related to UI or test cases please add screenshots, this will help reviewers and a reference will get created for future backtrack.

rt4914 reviewed (approved):
Great work @bladebunny .

Thanks a lot.
